server/server.go

```diff
@@ -68,15 +70,87 @@ func main() {
 		fmt.Fprintf(w, "web-api-gateway version: %s\nGo version: %s", version, runtime.Version())
 	})
 
-	logger := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	log.Printf("Starting web-api-gateway, version %s\n", version)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
 		log.Printf("Incoming Request %s %s %s", r.RemoteAddr, r.Method, r.URL)
 		http.DefaultServeMux.ServeHTTP(w, r)
 	})
 
-	log.Printf("Starting web-api-gateway, version %s\n", version)
-	log.Fatal(http.ListenAndServeTLS(*addr, *certFile, *keyFile, logger))
+	cr, err := NewCertificateReloader(*certFile, *keyFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	server := &http.Server{
+		Addr: ":https",
+		TLSConfig: &tls.Config{
+			GetCertificate: cr.GetCertificateFunc(),
+		},
+		Handler: mux,
+	}
+	log.Fatal(server.ListenAndServeTLS("", ""))
 }
 
+///////////////////////////////////////////////
+type certificateReloader struct {
+	certMu   sync.RWMutex
```

Comment:
Would it make sense to embed the mutex here?

Reply:
Thanks!